File: apps/web/src/shared/api/schemas.ts
import { z } from 'zod';

// --- Устройства ---
export const deviceSchema = z.object({
  id: z.number(),
  name: z.string(),
  samplerate: z.number(),
  channels: z.number(),
  default: z.boolean(),
});
export type Device = z.infer<typeof deviceSchema>;

export const devicesSchema = z.object({
  input: z.array(deviceSchema),
  output: z.array(deviceSchema),
});
export type Devices = z.infer<typeof devicesSchema>;

// --- Состояние движка ---
export const engineStateSchema = z.object({
  state: z.enum([
    'stopped',
    'loading',
    'listening',
    'recording',
    'thinking',
    'speaking',
    'paused',
    'error',
  ]),
});
export type EngineStateName = z.infer<typeof engineStateSchema>['state'];

// --- Настройки ---
export const audioSettingsSchema = z.object({
  input_device: z.number().nullable(),
  output_device: z.number().nullable(),
  input_gain: z.number(),
  output_gain: z.number(),
});

export const modelSettingsSchema = z.object({
  provider: z.string().nullable(),
  model: z.string().nullable(),
  triggers: z.array(z.string()),
});

export const settingsSchema = z.object({
  audio: audioSettingsSchema,
  assistant: z.object({
    name: z.string(),
    default_model: z.string(),
    language: z.string(),
  }),
  models: z.record(z.string(), modelSettingsSchema),
  tts: z.object({
    engine: z.string(),
    edge_voice: z.string(),
    edge_rate: z.string(),
  }),
  trigger: z.object({
    mode: z.string(),
    silence_seconds: z.number(),
    max_seconds: z.number(),
    beep: z.boolean(),
    wakeword: z.object({
      phrases: z.array(z.string()),
      fuzzy: z.number(),
    }),
  }),
  skills: z.object({ enabled: z.boolean() }),
});
export type Settings = z.infer<typeof settingsSchema>;

// --- Ключи (маскированно) ---
export const keysSchema = z.object({
  ANTHROPIC_API_KEY: z.string().nullable(),
  OPENAI_API_KEY: z.string().nullable(),
});
export type Keys = z.infer<typeof keysSchema>;

// --- Голоса TTS ---
export const voiceSchema = z.object({
  short_name: z.string(),
  gender: z.string(),
  name: z.string(),
});
export const voicesSchema = z.array(voiceSchema);
export type Voice = z.infer<typeof voiceSchema>;

// --- Скиллы ---
export const skillsSchema = z.object({
  enabled: z.boolean(),
  skills: z.array(z.object({ name: z.string() })),
});
export type SkillsInfo = z.infer<typeof skillsSchema>;

// --- События WebSocket ---
export const wsEventSchema = z.discriminatedUnion('type', [
  z.object({ type: z.literal('status'), state: engineStateSchema.shape.state }),
  z.object({
    type: z.literal('level'),
    source: z.enum(['input', 'output']),
    db: z.number(),
  }),
  z.object({
    type: z.literal('transcript'),
    role: z.enum(['user', 'assistant', 'system']),
    text: z.string(),
    model: z.string().optional(),
  }),
  z.object({ type: z.literal('error'), message: z.string() }),
]);
export type WsEvent = z.infer<typeof wsEventSchema>;

File: src/server/schemas.py
"""Pydantic-модели запросов. Зеркалят Zod-схемы фронта (shared/api)."""
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class AudioSettings(BaseModel):
    input_device: Optional[int] = None
    output_device: Optional[int] = None
    input_gain: float = Field(1.0, ge=0.0, le=2.0)
    output_gain: float = Field(1.0, ge=0.0, le=2.0)


class AssistantSettings(BaseModel):
    name: Optional[str] = None
    default_model: Optional[str] = None  # claude | gpt
    language: Optional[str] = None


class ModelSettings(BaseModel):
    model: Optional[str] = None


class TtsSettings(BaseModel):
    engine: Optional[str] = None         # edge | pyttsx3
    edge_voice: Optional[str] = None
    edge_rate: Optional[str] = None


class WakeSettings(BaseModel):
    phrases: Optional[list[str]] = None
    fuzzy: Optional[float] = Field(None, ge=0.0, le=1.0)


class TriggerSettings(BaseModel):
    mode: Optional[str] = None           # wakeword | push_to_talk
    end_mode: Optional[str] = None       # phrase | silence
    end_phrases: Optional[list[str]] = None
    silence_seconds: Optional[float] = None
    max_seconds: Optional[float] = None
    beep: Optional[bool] = None
    wakeword: Optional[WakeSettings] = None


class SkillsSettings(BaseModel):
    enabled: Optional[bool] = None


class SettingsUpdate(BaseModel):
    """Частичное обновление config.yaml. Передаются только изменённые секции."""
    audio: Optional[AudioSettings] = None
    assistant: Optional[AssistantSettings] = None
    models: Optional[dict[str, ModelSettings]] = None
    tts: Optional[TtsSettings] = None
    trigger: Optional[TriggerSettings] = None
    skills: Optional[SkillsSettings] = None


class KeysUpdate(BaseModel):
    """Новые значения ключей. Пустая строка/None — оставить без изменений."""
    ANTHROPIC_API_KEY: Optional[str] = None
    OPENAI_API_KEY: Optional[str] = None
    GEMINI_API_KEY: Optional[str] = None


class TtsTestRequest(BaseModel):
    text: str = "Привет! Это проверка голоса."
